### ISSUE TITLE
PHP fastcgi - failed (13: Permission denied) while reading upstream

### ISSUE THREAD

**Issue by devster31** [linuxserver/docker-nextcloud#40]
[linuxserverurl]: https://linuxserver.io
[![linuxserver.io](https://raw.githubusercontent.com/linuxserver/docker-templates/master/linuxserver.io/img/linuxserver_medium.png)][linuxserverurl]
I found the below permission issue with this container while trying to download a file:
```
2017/07/17 20:45:34 [crit] 305#305: *23 open() "/var/tmp/nginx/fastcgi/4/00/0000000004" failed (13: Permission denied) while reading upstream, client: <xx.xx.xx.xx>, server: _, request: "GET /<url> HTTP/1.1", upstream: "fastcgi://127.0.0.1:9000", host: "<host:port>"
```
The issue seems, as the error itself says, a permission issue, and therefore an issue with container permissions.
A similar problem was found in another container linuxserver/docker-muximux#6
Both the `php-fpm` process and the `nginx worker` process are running as `$PUID:$PGID`:
```
# ps -eo pid,user,group,comm,args
...
  283 root     root     php-fpm7         {php-fpm7} php-fpm: master process (/etc/php7/php-fpm.conf)
  286 root     root     nginx            nginx: master process /usr/sbin/nginx -c /config/nginx/nginx.conf
  293 abc      users    nginx            nginx: worker process
  294 abc      users    nginx            nginx: worker process
  295 abc      users    nginx            nginx: worker process
  296 abc      users    nginx            nginx: worker process
  306 abc      users    php-fpm7         {php-fpm7} php-fpm: pool www
  307 abc      users    php-fpm7         {php-fpm7} php-fpm: pool www
  308 abc      users    php-fpm7         {php-fpm7} php-fpm: pool www
...
```
The permission of the path that nginx uses as temp path are:
```
# ls -lah /var/tmp/nginx/
total 24K
drwx------ 6 abc  users 4.0K Jul 17 21:16 .
drwxr-xr-x 4 root root  4.0K Jul 17 21:16 ..
drwxr-xr-x 2 abc  users 4.0K Jul 17 21:16 client_body
drwx------ 2 abc  root  4.0K Jul 17 21:16 fastcgi
drwx------ 2 abc  root  4.0K Jul 17 21:16 proxy
drwx------ 2 abc  root  4.0K Jul 17 21:16 scgi
drwx------ 2 abc  root  4.0K Jul 17 21:16 uwsgi
```
There are a couple of questions on StackExchange that seem to solve this:
* [PHP FPM gives permission denied?](https://serverfault.com/questions/501258/php-fpm-gives-permission-denied)
* [Permission denied while reading upstream](https://serverfault.com/questions/235154/permission-denied-while-reading-upstream)

In particular it seems that the `<var>_temp_path` directories need specific permissions.
Below are the default nginx configure options on alpine:
```
--http-client-body-temp-path=/var/tmp/nginx/client_body
--http-proxy-temp-path=/var/tmp/nginx/proxy
--http-fastcgi-temp-path=/var/tmp/nginx/fastcgi
--http-uwsgi-temp-path=/var/tmp/nginx/uwsgi
--http-scgi-temp-path=/var/tmp/nginx/scgi
```
My personal fix, which is a bit of a hack, was to add the following line to the `/config/nginx/nginx.conf`:
```
    fastcgi_temp_path /tmp/fastcgi_temp 1 2;
```
Since `/tmp` is writable by all this seems to solve it.
##  Thanks, team linuxserver.io

**Comment by j0nnymoe** [comment]
Hi,
What's your host OS and also docker create command please?
Thanks

**Comment by devster31** [comment]
The command:
```shell
$ docker run -d \
    --name nextcloud \
    -p 4433:443 \
    -e PUID=`id -u` -e PGID=`id -g` \
    -v ${HOME}/nextcloud/data:/data \
    -v ${HOME}/nextcloud/config:/config \
    linuxserver/nextcloud
```
Below some system information:
```shell
$ docker -v
Docker version 17.06.0-ce, build 02c1d87
$ docker info
[...]
Server Version: 17.06.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 144
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 3.13.0-96-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.66GiB
[...]
```
However in my humble opinion it's extremely unlikely that this is caused by something outside of the container. The directory for which permission is denied isn't mounted or otherwise edited.

**Comment by aptalca** [comment]
We'll look into it, but most likely this is an nginx issue. The folder is correctly owned by the user abc (the user nginx is supposed to run as). That's all the container can do. 

If you google "nginx tmp permissions" you'll see a bunch of bug reports. It's an issue that pops up occasionally.

**Comment by aptalca** [comment]
I was able to reproduce similar errors with the letsencrypt container and wordpress (php fastcgi). 

After thorough testing I realized that with the same exact container and data, I get these errors no matter what on an ubuntu 14 host but not on Ubuntu 16. Even tested different versions of docker but at the end, the only difference was that it didn't work properly on the Ubuntu 14 host. Not sure what to tell you. There is absolutely nothing wrong with the permissions inside the image. All this tmp files are owned by the user abc inside the container (which takes the id of the puid variable) and both nginx and php-fpm run as abc.

It's great to hear that your solution works. Because without it, my only suggestion would be to upgrade the host to Ubuntu 16
